src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSynchronizer.scala

```diff
@@ -49,90 +47,132 @@ class GoogleGroupSynchronizer(
-        removeTrials <- Future.traverse(toRemove) { removeEmail =>
-          googleDirectoryDAO.removeMemberFromGroup(group.email, WorkbenchEmail(removeEmail)).toTry.map(toSyncReportItem("removed", removeEmail, _))
-        }
+        // use proxy group email instead of user's actual email
+        case userSubjectId: WorkbenchUserId => googleExtensions.getUserProxy(userSubjectId)
 
-        _ <- directoryDAO.updateSynchronizedDate(groupId, samRequestContext).unsafeToFuture()
-      } yield Map(group.email -> Seq(addTrials, removeTrials).flatten) ++ subGroupSyncs.flatten
+        // not sure why this next case would happen but if a petSA is in a group just use its email
+        case petSA: PetServiceAccountId => directoryDAO.loadSubjectEmail(petSA, samRequestContext)
+      }
+      .map(_.collect { case Some(email) => email.value.toLowerCase }.toSet)
+
+  /** If the group exists in google, return its members, otherwise create it and return empty
+    * @param group
+    * @return
+    */
+  private def loadGoogleGroupMemberEmailsMaybeCreateGroup(group: WorkbenchGroup, samRequestContext: SamRequestContext) = {
+    // note that these 2 definitions are IOs so they don't actually do anything until they are used below
+    val listMembersIO = traceIOWithContext("listGoogleGroupMembers", samRequestContext) { _ =>
+      IO.fromFuture(IO(googleDirectoryDAO.listGroupMembers(group.email)))
+    }
+    val createGroupIO = traceIOWithContext("createGoogleGroup", samRequestContext) { _ =>
+      IO.fromFuture(IO(googleDirectoryDAO.createGroup(group.id.toString, group.email, Option(googleDirectoryDAO.lockedDownGroupSettings))))
+    }
+
+    listMembersIO flatMap {
+      case None => createGroupIO map (_ => Set.empty[String])
+      case Some(members) => IO.pure(members.map(_.toLowerCase).toSet)
     }
   }
 
+  /** synchronize any sub groups that have never been synchronized otherwise they don't exist in google and can't be added
+    * @param group
+    * @param visitedGroups
+    * @param samRequestContext
+    * @return
+    */
+  private def maybeSyncSubGroups(group: WorkbenchGroup, visitedGroups: Set[WorkbenchGroupIdentity], samRequestContext: SamRequestContext) =
+    group.members.toList.traverse {
+      case subGroup: WorkbenchGroupIdentity =>
+        directoryDAO.getSynchronizedDate(subGroup, samRequestContext).flatMap {
+          case None => synchronizeGroupMembers(subGroup, visitedGroups + group.id, samRequestContext)
+          case _ => IO.pure(Map.empty[WorkbenchEmail, Seq[SyncReportItem]])
+        }
+      case _ => IO.pure(Map.empty[WorkbenchEmail, Seq[SyncReportItem]])
+    }
+
+  /** If this is a policy group constrained by an auth domain calculate the intersection group membership, otherwise just return the group membership
+    * @param group
+    * @param samRequestContext
+    * @return
+    */
+  private def maybeCalculateIntersection(group: WorkbenchGroup, samRequestContext: SamRequestContext): IO[Set[WorkbenchSubject]] =
```

Comment:
in my head, `maybe` in a variable name is an `Option`, but `maybe` in a function name means it there is some conditionality but I understand overloading is not optimal.

how about instead of prefix with `maybe`, suffix with `IfRequired`?